File: examples/python/framing_and_session_example.py
"""Framing, session handshake and checksum helpers.

The schema-driven examples cover encode/decode. This example rounds out the
remaining runtime surface that interoperates with the C++ ``FrameChannel`` and
``UprSession``:

* length-prefixed framing: ``encode_frame``, ``try_read_frame``, ``iter_frames``
  and the stateful ``FrameDecoder``;
* the ``UPR1`` session handshake: ``encode_handshake`` / ``decode_handshake`` and
  ``check_compatibility``;
* the built-in checksum algorithms.

None of this needs a schema - it operates on raw payload bytes - so it applies
to any protocol you encode with the ``Codec``.
"""

from __future__ import annotations

from universal_protocol_runtime import (
    FrameDecoder,
    FramingError,
    Handshake,
    TransportMode,
    check_compatibility,
    checksums,
    decode_handshake,
    encode_frame,
    encode_handshake,
    iter_frames,
    try_read_frame,
)


def demo_framing() -> None:
    print("== framing ==")
    payloads = [b"alpha", b"bravo", b"charlie"]

    # Default prefix width is 4 bytes (little-endian). Use prefix_width=1/2 to
    # match a tighter wire format such as the hardware stream example.
    wire = b"".join(encode_frame(p) for p in payloads)
    print(f"wire_bytes={len(wire)} first_frame_hex={encode_frame(payloads[0]).hex()}")

    # One-shot: pull every complete frame out of a fully-buffered blob.
    print(f"iter_frames={[bytes(f) for f in iter_frames(wire)]}")

    # Single-frame parse returning (payload, bytes_consumed), or None if the
    # buffer does not yet hold a whole frame.
    head = try_read_frame(wire)
    assert head is not None
    payload, consumed = head
    print(f"try_read_frame payload={bytes(payload)!r} consumed={consumed}")

    # Streaming: a FrameDecoder reassembles frames from arbitrary chunk
    # boundaries (here we split mid-frame on purpose).
    decoder = FrameDecoder()
    received: list[bytes] = []
    for index in range(0, len(wire), 3):
        received.extend(decoder.feed(wire[index:index + 3]))
    print(f"frame_decoder={received}")
    assert received == payloads


def demo_handshake() -> None:
    print("== session handshake ==")
    local = Handshake(
        transport_mode=TransportMode.LENGTH_PREFIXED_STREAM,
        max_frame_bytes=1 << 20,
        session_id=0x1122334455667788,
    )
    blob = encode_handshake(local)
    print(f"handshake_bytes={len(blob)} hex={blob.hex()}")

    parsed = decode_handshake(blob)
    assert parsed == local
    print(f"decoded transport_mode={parsed.transport_mode.name} session_id={parsed.session_id:#x}")

    # A peer that advertises a smaller max frame size is rejected.
    remote_ok = Handshake(max_frame_bytes=1 << 20, session_id=0x99)
    check_compatibility(local, remote_ok)
    print("compatible_peer=ok")

    remote_bad = Handshake(max_frame_bytes=1024, session_id=0x99)
    try:
        check_compatibility(local, remote_bad)
    except FramingError as exc:
        print(f"incompatible_peer rejected: {exc}")


def demo_checksums() -> None:
    print("== checksums ==")
    data = b"123456789"  # the canonical checksum test vector
    for algorithm in ("xor8", "sum16", "crc16_ccitt", "crc32", "crc32c"):
        digest = checksums.compute(algorithm, memoryview(data))
        print(f"{algorithm:>12} = {digest:#010x}")


def main() -> int:
    demo_framing()
    demo_handshake()
    demo_checksums()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/typescript/framing_and_session_example.ts
/**
 * Use runtime helpers that operate on raw wire bytes.
 *
 * Shows length-prefixed framing, streaming frame accumulation, UPR1 handshake
 * encoding/compatibility checks, and the built-in checksum algorithms.
 */

import {
  FrameDecoder,
  FramingError,
  TransportMode,
  checkCompatibility,
  checksums,
  decodeHandshake,
  defaultHandshake,
  encodeFrame,
  encodeHandshake,
  iterFrames,
  tryReadFrame,
} from "universal-protocol-runtime";

const TEXT = new TextEncoder();

function toHex(bytes: Uint8Array): string {
  return Array.from(bytes, (b) => b.toString(16).padStart(2, "0")).join("");
}

function decode(bytes: Uint8Array): string {
  return new TextDecoder().decode(bytes);
}

function concat(parts: Uint8Array[]): Uint8Array {
  const total = parts.reduce((sum, p) => sum + p.length, 0);
  const out = new Uint8Array(total);
  let offset = 0;
  for (const part of parts) {
    out.set(part, offset);
    offset += part.length;
  }
  return out;
}

function demoFraming(): void {
  console.log("== framing ==");
  const payloads = ["alpha", "bravo", "charlie"].map((s) => TEXT.encode(s));

  // Default prefix width is 4 bytes (little-endian). Use prefixWidth 1/2 to match
  // a tighter wire format such as the hardware stream example.
  const wire = concat(payloads.map((p) => encodeFrame(p)));
  console.log(`wire_bytes=${wire.length} first_frame_hex=${toHex(encodeFrame(payloads[0]))}`);

  // One-shot: pull every complete frame out of a fully-buffered blob.
  console.log(`iter_frames=${[...iterFrames(wire)].map(decode).join(",")}`);

  // Single-frame parse returning { payload, consumed }, or null if the buffer
  // does not yet hold a whole frame.
  const head = tryReadFrame(wire);
  if (head === null) {
    throw new Error("expected a complete frame");
  }
  console.log(`try_read_frame payload=${decode(head.payload)} consumed=${head.consumed}`);

  // Streaming: a FrameDecoder reassembles frames from arbitrary chunk boundaries
  // (here we split mid-frame on purpose).
  const decoder = new FrameDecoder();
  const received: string[] = [];
  for (let index = 0; index < wire.length; index += 3) {
    for (const frame of decoder.feed(wire.subarray(index, index + 3))) {
      received.push(decode(frame));
    }
  }
  console.log(`frame_decoder=${received.join(",")}`);
}

function demoHandshake(): void {
  console.log("== session handshake ==");
  const local = defaultHandshake({
    transportMode: TransportMode.LengthPrefixedStream,
    maxFrameBytes: 1 << 20,
    sessionId: 0x1122334455667788n,
  });
  const blob = encodeHandshake(local);
  console.log(`handshake_bytes=${blob.length} hex=${toHex(blob)}`);

  const parsed = decodeHandshake(blob);
  console.log(`decoded transport_mode=${parsed.transportMode} session_id=0x${parsed.sessionId.toString(16)}`);

  // A peer that advertises a smaller max frame size is rejected.
  checkCompatibility(local, defaultHandshake({ maxFrameBytes: 1 << 20 }));
  console.log("compatible_peer=ok");

  try {
    checkCompatibility(local, defaultHandshake({ maxFrameBytes: 1024 }));
  } catch (error) {
    if (error instanceof FramingError) {
      console.log(`incompatible_peer rejected: ${error.message}`);
    } else {
      throw error;
    }
  }
}

function demoChecksums(): void {
  console.log("== checksums ==");
  const data = TEXT.encode("123456789"); // the canonical checksum test vector
  for (const algorithm of ["xor8", "sum16", "crc16_ccitt", "crc32", "crc32c"]) {
    const digest = checksums.compute(algorithm, data);
    console.log(`${algorithm.padStart(12)} = 0x${(digest >>> 0).toString(16).padStart(8, "0")}`);
  }
}

function main(): void {
  demoFraming();
  demoHandshake();
  demoChecksums();
}

main();
